pkg/servicecache/service_cache.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2020 Authors of Hubble
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package servicecache
+
+import (
+	"fmt"
+	"net"
+	"strconv"
+	"sync"
+
+	pb "github.com/cilium/hubble/api/v1/flow"
+
+	"github.com/cilium/cilium/api/v1/models"
+)
+
+// ServiceCache is a cache of existing services.
+type ServiceCache struct {
+	mu    sync.RWMutex
+	cache map[string]*entry
+}
+
+// New creates a new empty ServiceCache.
+func New() *ServiceCache {
+	return &ServiceCache{
+		cache: map[string]*entry{},
+	}
+}
+
+type entry struct {
```

Comment:
why does this un-exported struct have a bunch of exported members? usually that is only a pattern for (un)marshaling data from other formats (reflect requires public vars)